packages/translator/src/Charcoal/Translator/Translator.php

```diff
@@ -204,21 +208,37 @@ public function translationChoice($val, $number, array $parameters = [], $domain
             return null;
         }
 
+        // Convert old parameters
+        $oldParameters = $parameters;
+        $parameters = [];
+        foreach ($oldParameters as $key => $value) {
+            $key = preg_replace('/^%(\w+)%$/', '$1', $key);
+            $parameters[$key] = $value;
+        }
+
         $parameters = array_merge([
-            '%count%' => $number,
+            'count' => $number,
         ], $parameters);
 
         $translation = new Translation($val, $this->manager());
         $localized   = (string)$translation;
+
         foreach ($this->availableLocales() as $lang) {
+            $hasTranslation = $this->hasTrans($localized, ($domain ?? null), $lang);
             if (!isset($translation[$lang]) || $translation[$lang] === $val) {
-                $translation[$lang] = $this->transChoice($localized, $number, $parameters, $domain, $lang);
+                if ($hasTranslation) {
+                    $translation[$lang] = $this->translationRaw($localized, ($domain ?? null))[$lang];
+                    $translation[$lang] = $this->convertLegacyChoiceFormat((string)$translation[$lang], $parameters);
+                } elseif (isset($translation[$lang])) {
+                    $translation[$lang] = $this->convertLegacyChoiceFormat((string)$val, $parameters);
+                } else {
+                    continue;
+                }
             } else {
-                $translation[$lang] = strtr(
-                    $this->selector()->choose($translation[$lang], (int)$number, $lang),
-                    $parameters
-                );
+                $translation[$lang] = $this->convertLegacyChoiceFormat((string)$translation[$lang], $parameters);
             }
+
+            $translation[$lang] = $this->formatMessage($lang, $translation[$lang], $parameters);
         }
```

Comment:
The logic of this iteration feels complicated (especially with the dubious usage of `translationRaw()`.

```suggestion
        foreach ($this->availableLocales() as $lang) {
            if (!isset($translation[$lang])) {
                if (!$this->hasTrans($localized, $domain, $lang)) {
                    continue;
                }

                $translation[$lang] = $this->trans($localized, [], $domain, $lang);
            }

            $translation[$lang] = $this->convertLegacyChoiceFormat($translation[$lang], $parameters);
            $translation[$lang] = $this->formatMessage($lang, $translation[$lang], $parameters);
        }
```